File: app/src/main/java/com/appharbr/example/app/admob/AdMobNativeAdActivity.java
/*
 * Copyright (C) 2015 Google, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.appharbr.example.app.admob;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.RatingBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.appharbr.sdk.engine.AdResult;
import com.appharbr.sdk.engine.AdSdk;
import com.appharbr.sdk.engine.AdStateResult;
import com.appharbr.sdk.engine.AppHarbr;
import com.appharbr.example.app.R;
import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdLoader;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.LoadAdError;
import com.google.android.gms.ads.VideoController;
import com.google.android.gms.ads.VideoOptions;
import com.google.android.gms.ads.formats.NativeAdOptions;
import com.google.android.gms.ads.nativead.NativeAd;
import com.google.android.gms.ads.nativead.NativeAdView;

import java.util.Locale;

/**
 * A simple activity class that displays native ad formats.
 */
public class AdMobNativeAdActivity extends AppCompatActivity {

    private Button refresh;
    private CheckBox startVideoAdsMuted;
    private TextView videoStatus;
    private NativeAd adMobNativeAd;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.admob_native_activity);
        refresh = findViewById(R.id.btn_refresh);
        startVideoAdsMuted = findViewById(R.id.cb_start_muted);
        videoStatus = findViewById(R.id.tv_video_status);
        refresh.setOnClickListener(unusedView -> requestAd());

        requestAd();
    }

    private void requestAd() {
        refresh.setEnabled(false);

        AdLoader.Builder builder = new AdLoader.Builder(this, getApplicationContext().getResources().getString(R.string.admob_native_ad_unit_id));

        builder.forNativeAd(nativeAd -> {
            if (isDestroyed()) {
                nativeAd.destroy();
                return;
            }

            if (adMobNativeAd != null) {
                adMobNativeAd.destroy();
            }
            adMobNativeAd = nativeAd;
            FrameLayout frameLayout = findViewById(R.id.fl_adplaceholder);
            NativeAdView adView = (NativeAdView) getLayoutInflater().inflate(R.layout.admob_ad_unified, null);
            populateUnifiedNativeAdView(adMobNativeAd, adView);
            frameLayout.removeAllViews();
            frameLayout.addView(adView);
        });

        VideoOptions videoOptions = new VideoOptions.Builder()
                .setStartMuted(startVideoAdsMuted.isChecked())
                .build();

        NativeAdOptions adOptions = new NativeAdOptions.Builder()
                .setVideoOptions(videoOptions)
                .build();

        builder.withNativeAdOptions(adOptions);

        AdLoader adLoader =
                builder
                        .withAdListener(
                                new AdListener() {
                                    @Override
                                    public void onAdFailedToLoad(@NonNull LoadAdError loadAdError) {
                                        refresh.setEnabled(true);
                                    }
                                })
                        .build();

        adLoader.loadAd(new AdRequest.Builder().build());

        videoStatus.setText("");
    }


    private void populateUnifiedNativeAdView(NativeAd nativeAd, NativeAdView adView) {
        // ############ Check with AppHarbr if this ad suppose to be display ###########
        AdResult adResult = AppHarbr.shouldBlockNativeAd(AdSdk.ADMOB, nativeAd);
        if (adResult.getAdStateResult() == AdStateResult.BLOCKED) {
            // **** AppHarbr Blocked The Native Ad - Do Not Render it. ****
            // May request another ad.
            refresh.setEnabled(true);
            return;
        }

        adView.setMediaView(adView.findViewById(R.id.ad_media));
        adView.setHeadlineView(adView.findViewById(R.id.ad_headline));
        adView.setBodyView(adView.findViewById(R.id.ad_body));
        adView.setCallToActionView(adView.findViewById(R.id.ad_call_to_action));
        adView.setIconView(adView.findViewById(R.id.ad_app_icon));
        adView.setPriceView(adView.findViewById(R.id.ad_price));
        adView.setStarRatingView(adView.findViewById(R.id.ad_stars));
        adView.setStoreView(adView.findViewById(R.id.ad_store));
        adView.setAdvertiserView(adView.findViewById(R.id.ad_advertiser));
        ((TextView) adView.getHeadlineView()).setText(nativeAd.getHeadline());
        adView.getMediaView().setMediaContent(nativeAd.getMediaContent());
        if (nativeAd.getBody() == null) {
          adView.getBodyView().setVisibility(View.INVISIBLE);
        } else {
          adView.getBodyView().setVisibility(View.VISIBLE);
          ((TextView) adView.getBodyView()).setText(nativeAd.getBody());
        }
        if (nativeAd.getCallToAction() == null) {
          adView.getCallToActionView().setVisibility(View.INVISIBLE);
        } else {
          adView.getCallToActionView().setVisibility(View.VISIBLE);
          ((Button) adView.getCallToActionView()).setText(nativeAd.getCallToAction());
        }
        if (nativeAd.getIcon() == null) {
            adView.getIconView().setVisibility(View.GONE);
        } else {
            ((ImageView) adView.getIconView()).setImageDrawable(
                    nativeAd.getIcon().getDrawable());
            adView.getIconView().setVisibility(View.VISIBLE);
        }
        if (nativeAd.getPrice() == null) {
            adView.getPriceView().setVisibility(View.INVISIBLE);
        } else {
            adView.getPriceView().setVisibility(View.VISIBLE);
            ((TextView) adView.getPriceView()).setText(nativeAd.getPrice());
        }
        if (nativeAd.getStore() == null) {
            adView.getStoreView().setVisibility(View.INVISIBLE);
        } else {
            adView.getStoreView().setVisibility(View.VISIBLE);
            ((TextView) adView.getStoreView()).setText(nativeAd.getStore());
        }
        if (nativeAd.getStarRating() == null) {
            adView.getStarRatingView().setVisibility(View.INVISIBLE);
        } else {
            ((RatingBar) adView.getStarRatingView())
                    .setRating(nativeAd.getStarRating().floatValue());
            adView.getStarRatingView().setVisibility(View.VISIBLE);
        }
        if (nativeAd.getAdvertiser() == null) {
            adView.getAdvertiserView().setVisibility(View.INVISIBLE);
        } else {
            ((TextView) adView.getAdvertiserView()).setText(nativeAd.getAdvertiser());
            adView.getAdvertiserView().setVisibility(View.VISIBLE);
        }
        adView.setNativeAd(nativeAd);
        VideoController vc = nativeAd.getMediaContent().getVideoController();
        if (vc.hasVideoContent()) {
            videoStatus.setText(String.format(Locale.getDefault(),
                    "Video status: Ad contains video asset."));
            vc.setVideoLifecycleCallbacks(new VideoController.VideoLifecycleCallbacks() {
                @Override
                public void onVideoEnd() {
                    refresh.setEnabled(true);
                    videoStatus.setText("Video status: Video playback has ended.");
                    super.onVideoEnd();
                }
            });
        } else {
            videoStatus.setText("Video status: Ad does not contain a video asset.");
            refresh.setEnabled(true);
        }
    }

    @Override
    protected void onDestroy() {
        if (adMobNativeAd != null) {
            adMobNativeAd.destroy();
        }
        super.onDestroy();
    }
}

File: kotlinsampleapp/src/main/java/com/appharbr/kotlin/example/app/admob/AdmobNativeAdActivity.kt
package com.appharbr.kotlin.example.app.admob

import android.os.Bundle
import android.util.Log
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.material.*
import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateOf
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.sp
import androidx.core.graphics.drawable.toBitmap
import com.appharbr.kotlin.example.app.R
import com.appharbr.kotlin.example.app.ui.theme.AppHarbrExampleAppTheme
import com.appharbr.sdk.engine.AdSdk
import com.appharbr.sdk.engine.AdStateResult
import com.appharbr.sdk.engine.AppHarbr
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdLoader
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.nativead.NativeAd
import com.google.android.gms.ads.nativead.NativeAdOptions

class AdmobNativeAdActivity : ComponentActivity() {

    private val nativeAdState = mutableStateOf<NativeAd?>(null)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        requestNativeAd()

        setContent {
            AppHarbrExampleAppTheme {
                Surface(
                    modifier = Modifier.fillMaxSize(),
                    color = MaterialTheme.colors.background
                ) {
                    Column(
                        modifier = Modifier.fillMaxSize(),
                        verticalArrangement = Arrangement.Center,
                        horizontalAlignment = Alignment.CenterHorizontally
                    ) {
                        DisplayNativeAd()
                    }
                }
            }
        }
    }

    private fun requestNativeAd() {
        val adLoader = AdLoader.Builder(
            this,
            applicationContext.resources.getString(R.string.admob_native_ad_unit_id)
        ).forNativeAd { nativeAd: NativeAd ->

            val adResult = AppHarbr.shouldBlockNativeAd(AdSdk.ADMOB, nativeAd, applicationContext.resources.getString(R.string.admob_native_ad_unit_id))
            when (adResult.adStateResult) {
                AdStateResult.BLOCKED -> {
                    nativeAdState.value = null
                    Log.e("LOG", "Native ad was blocked by appharbr")
                }
                else -> {}
            }

            if (isDestroyed) {
                nativeAdState.value?.destroy()
                return@forNativeAd
            }
            nativeAdState.value?.destroy()
            nativeAdState.value = nativeAd
        }.withAdListener(object : AdListener() {
            override fun onAdFailedToLoad(adError: LoadAdError) {
                Log.e("LOG", "Handle the failure by logging, altering the UI, and so on.")
            }
        }).withNativeAdOptions(
            NativeAdOptions.Builder().build()
        ).build()

        adLoader.loadAd(AdRequest.Builder().build())
    }

    @Composable
    private fun DisplayNativeAd() {
        nativeAdState.value?.let {
            Text(text = "Headline: ${it.headline ?: "Empty Headline"}")
            Text(text = "Body: ${it.body ?: "Empty body"}")
            it.icon?.drawable?.toBitmap()?.let { bitmap -> Image(bitmap.asImageBitmap(), "") }
            for (image in it.images) {
                image.drawable?.toBitmap()?.let { bitmap -> Image(bitmap.asImageBitmap(), "") }
            }
        } ?: kotlin.run {
            Text(
                text = stringResource(id = R.string.admob_native_screen),
                fontSize = 20.sp
            )
            CircularProgressIndicator()
        }
    }

}